core/src/main/java/de/javakaffee/web/msm/storage/RedisStorageClient.java

```diff
@@ -0,0 +1,292 @@
+                        _log.debug(String.format("Creating new Jedis instance (host=%s, port=%s, ssl=%s)",
+                            _host, _port, _ssl));
+                    }
+
+                    return createJedisInstance();
+                } else {
+                    if (knownGood) {
+                        // Check all existing connections until we find a good one
+                        BinaryJedis jedis;
+                        do {
+                            jedis = _instances.remove(_instances.size() - 1);
+                            try {
+                                jedis.ping();
+                                
+                                if (_log.isDebugEnabled())
+                                    _log.debug(String.format("Using known-good connection #%d", _instances.size()));
+
+                                return jedis;
+                            } catch (Exception e) {
+                                if (_log.isDebugEnabled()) {
+                                    _log.debug(String.format("Removing connection #%d since it cannot be pinged", _instances.size()));
+                                }
+                            }
+                        } while (!_instances.isEmpty());
+
+                        // No existing connections are good, so create new one
+                        if (_log.isDebugEnabled()) {
+                            _log.debug(String.format("Creating new Jedis instance (host=%s, port=%s, ssl=%s) since all existing connections were bad",
+                                _host, _port, _ssl));
+                        }
+                        
+                        return createJedisInstance();
+                    } else {
+                        if (_log.isDebugEnabled())
+                            _log.debug(String.format("Using connection #%d", _instances.size() - 1));
+
+                        return _instances.remove(_instances.size() - 1);
+                    }
+                }
+            }
+        }
+        
+        public void returnInstance(BinaryJedis instance) {
+            synchronized (_instances) {
+                _instances.add(instance);
+                
+                if (_log.isDebugEnabled())
+                    _log.debug(String.format("Returned instance #%d", _instances.size() - 1));
+            }
+        }
+        
+        public void shutdown() {
+            synchronized (_instances) {
+                if (_log.isDebugEnabled())
+                    _log.debug(String.format("Closing %d remaining Jedis instance(s)", _instances.size()));
+                
+                for (BinaryJedis jedis: _instances) {
+                    try { jedis.close(); } catch (Exception e) { /* ignore exception */ }
+                }
+                _instances.clear();
+            }
+        }
+        
+        private BinaryJedis createJedisInstance() {
+            return new BinaryJedis(_host, _port, _ssl);
```

Comment:
A suitable timeout could/should be configured, perhaps the configured msm operationTimeout.